### PR TITLE
Kraken: Implementation of cached next_stop_time

### PR DESCRIPTION
**Pull request by xlqian**
http://jira.canaltp.fr/browse/NAVP-422 (+ http://jira.canaltp.fr/browse/NAVP-480)
- Implementation of Load cache
- NOT implemented LRU

Benchmarked on Stif
- 180ms per request without cache init
-  250ms per request with cache init (for each request)

[Review on Reviewable](https://reviewable.io/reviews/canaltp/navitia/1406)

Result of bench:
- 100 requests, every request is repeated for 10 times
- Benched on BackupStif, fr-pdl, SNCF
- Benched on `dev` (right before this PR, contributor is only activated on VJ), `Cache`, `LRU`(to do)

**Backup STIF** : 943 solutions / 1000 requests

| (nb 2nd pass) | 0 | 10 | 25 | 100 | 1000 | 10000 |
| --- | --- | --- | --- | --- | --- | --- |
| nb journeys found | 6682 | 6930 | 7098 | 7603 | 7715 | 7715 |
| dev | 125.76s | 226.98s | 354.58 | 715.26s | 885.6s | 873s |
| Cache | 219.41s | 260s | 311.56s | 461.82s | 517.49s | 552.78s |
| LRU | 78.88s | 138.88s | 189.54s | 399s | 482s | 475s |

**fr-pdl**:  863 solutions / 1000 requests

| (nb 2nd pass) | 0 | 10 | 100 | 1000 | 10000 |
| --- | --- | --- | --- | --- | --- |
| dev | 78.8s | 128s | 251s | 280s | 281s |
| Cache | 120s | 126.13s | 193.09s | 203s | 204s |
| LRU | 60s | 90s | 162s | 180s | 181s |

**SNCF**:  1000 solutions / 1000 requests

| (nb 2nd pass) | 0 | 10 | 100 | 1000 | 10000 |
| --- | --- | --- | --- | --- | --- |
| dev | 11.62s | 12.37s | 13.17s | 12.36s | 12.38s |
| Cache | 15.7s | 16.01 | 16.17s | 16.73s | 16.69s |
| LRU | 7.28s | 7.48s | 7.51s | 7.56s | 8.2s |

### CODE REVIEW

File: source/routing/benchmark_full.cpp [hunk 1]
```diff
@@ -281,14 +282,15 @@ int main(int argc, char** argv){
...
         auto resp = make_response(router, origin, destination,
-              {DateTimeUtils::set(date.days(), demand.hour)}, true,
+              {departure_datetime}, true,
               accessibilite_params,
               {},
               georef_worker,
               type::RTLevel::Base,
               2_min,
-              std::numeric_limits<int>::max(),
+              departure_datetime + DateTimeUtils::SECONDS_PER_DAY,
```
Comment (TeXitoi) [inline source/routing/benchmark_full.cpp]: It should be only `DateTimeUtils::SECONDS_PER_DAY`

**Comment by xlqian** [comment]
Jenkins, test this please!

File: source/routing/next_stop_time.cpp [hunk 2]
```diff
@@ -338,6 +341,126 @@ NextStopTime::tardiest_stop_time(const StopEvent stop_event,
...
+        const type::RTLevel rt_level,
+        const type::AccessibiliteParams& accessibilite_params,
+        const JourneyPattern& jp,
+        const std::vector<const VJ_T*>& vjs,
+        IdxMap<JourneyPatternPoint, std::vector<DtSt>>& arrival,
+        IdxMap<JourneyPatternPoint, std::vector<DtSt>>& departure){
+
+    const int to_int = static_cast<int>(DateTimeUtils::date(to));
+    int from_int = static_cast<int>(DateTimeUtils::date(from)) - 1;
+    from_int = from_int > 0 ? from_int : 0;
```
Comment (antoine-de) [inline source/routing/next_stop_time.cpp]: `from_int = std::max(from_int, 0);` ?

File: source/routing/next_stop_time.cpp [hunk 3]
```diff
@@ -338,6 +341,126 @@ NextStopTime::tardiest_stop_time(const StopEvent stop_event,
...
+
+    for (const auto* vj :  vjs) {
```
Comment (antoine-de) [inline source/routing/next_stop_time.cpp]: spacing

File: source/routing/next_stop_time.cpp [hunk 4]
```diff
@@ -338,6 +341,126 @@ NextStopTime::tardiest_stop_time(const StopEvent stop_event,
...
+        const DateTime to,
+        const type::RTLevel rt_level,
+        const type::AccessibiliteParams& accessibilite_params,
+        const JourneyPattern& jp,
+        const std::vector<const VJ_T*>& vjs,
+        IdxMap<JourneyPatternPoint, std::vector<DtSt>>& arrival,
+        IdxMap<JourneyPatternPoint, std::vector<DtSt>>& departure){
+
+    const int to_int = static_cast<int>(DateTimeUtils::date(to));
+    int from_int = static_cast<int>(DateTimeUtils::date(from)) - 1;
```
Comment (antoine-de) [inline source/routing/next_stop_time.cpp]: maybe add a comment that you want the day before too no ?

File: source/routing/next_stop_time.cpp [hunk 5]
```diff
@@ -338,6 +341,126 @@ NextStopTime::tardiest_stop_time(const StopEvent stop_event,
...
+    from_int = from_int > 0 ? from_int : 0;
+
+    for (const auto* vj :  vjs) {
+        if (! vj->accessible(accessibilite_params.vehicle_properties)) {
+            continue;
+        }
+        // test validity pattern of vj
+        const auto* vp = vj->validity_patterns[rt_level];
+        for (int day = from_int; day <=  to_int ; ++day) {
```
Comment (antoine-de) [inline source/routing/next_stop_time.cpp]: spacing

File: source/routing/next_stop_time.cpp [hunk 6]
```diff
@@ -338,6 +341,126 @@ NextStopTime::tardiest_stop_time(const StopEvent stop_event,
...
+            if (! vp->check(day)) {
+                continue;
+            }
+            const auto shift = navitia::DateTimeUtils::SECONDS_PER_DAY * day;
+            size_t i= 0;
```
Comment (antoine-de) [inline source/routing/next_stop_time.cpp]: spacing :wink:

File: source/routing/next_stop_time.cpp [hunk 7]
```diff
@@ -338,6 +341,126 @@ NextStopTime::tardiest_stop_time(const StopEvent stop_event,
...
+template<typename F>
+static void vj_loop(const nt::DiscreteVehicleJourney*, F f) {
+    // Discrete Vehicle Journey doesn't need to loop
+    f(0);
+}
+
+template<typename F>
+static void vj_loop(const nt::FrequencyVehicleJourney* vj, F f) {
+    for (auto freq_shift = vj->start_time; freq_shift <= vj->end_time; freq_shift+= vj->headway_secs) {
+        f(freq_shift);
```
Comment (antoine-de) [inline source/routing/next_stop_time.cpp:368]: so now we expand the frequency vj ?

We need to bench this on datasets with ODT to check the differences.

File: source/routing/next_stop_time.h [hunk 8]
```diff
@@ -195,6 +197,30 @@ struct NextStopTime {
...
+    // either a vehicle that leaves or that arrives depending on
+    // clockwise.
+    std::pair<const type::StopTime*, DateTime>
+    next_stop_time(const StopEvent stop_event,
+                   const JppIdx jpp_idx,
+                   const DateTime dt,
+                   const bool clockwise) const;
+
+private:
+    using DtSt = std::pair<DateTime, const type::StopTime*>;
```
Comment (antoine-de) [inline source/routing/next_stop_time.h]: there is several time this alias, can't we share it ?

File: source/routing/tests/raptor_test.cpp [hunk 9]
```diff
@@ -1951,7 +1953,7 @@ BOOST_AUTO_TEST_CASE(good_connection_when_walking_as_fast_as_bus) {
     auto res1 = raptor.compute(d.stop_areas_map.at("A"), d.stop_areas_map.at("E"),
                                "8:00"_t, 0, DateTimeUtils::inf, type::RTLevel::Base, 2_min, true);
     test_good_connection_when_walking_as_fast_as_bus(res1);
-
+    std::cout << "-----------------ok-----------------" << std::endl;
```
Comment (antoine-de) [inline source/routing/tests/raptor_test.cpp]: debug no ?

File: source/routing/tests/raptor_test.cpp [hunk 10]
```diff
@@ -2044,7 +2046,7 @@ BOOST_AUTO_TEST_CASE(direct_path_filter) {
     // reverse clockwise
     res = raptor.compute_all(departures,
                              arrivals,
-                             DateTimeUtils::set(2, "08:03"_t),
+                             DateTimeUtils::set(2, "08:05"_t),
```
Comment (antoine-de) [inline source/routing/tests/raptor_test.cpp:2048]: why did you need to change those ?

File: source/routing/tests/routing_api_test.cpp [hunk 11]
```diff
@@ -1293,6 +1293,7 @@ BOOST_FIXTURE_TEST_CASE(bus_car_parking, streetnetworkmode_fixture<test_speed_pr
         seconds(distance_ar / get_default_speed()[Mode_e::Walking] / speed_factor)
         + seconds(1);
     destination.streetnetwork_params.speed_factor = speed_factor;
+    datetimes = {navitia::test::to_posix_timestamp("20120614T070000")};
```
Comment (antoine-de) [inline source/routing/tests/routing_api_test.cpp:1296]: why did you need to add this ?

File: source/routing/benchmark_full.cpp [hunk 12]
```diff
@@ -125,7 +125,7 @@ static type::EntryPoint make_entry_point(const std::string& entry_id, const type
     try {
         type::idx_t idx = boost::lexical_cast<type::idx_t>(entry_id);
 
-        //if it is a idx, we consider it to be a stop area idx
+        //if it iCached s a idx, we consider it to be a stop area idx
```
Comment (pbougue) [inline source/routing/benchmark_full.cpp]: miss-spelled

File: source/routing/next_stop_time.cpp [hunk 13]
```diff
@@ -52,8 +55,8 @@ bool NextStopTimeData::Arrival::is_valid(const type::StopTime& st) const {
 
 template<typename Getter>
 void NextStopTimeData::TimesStopTimes<Getter>::init(const JourneyPattern& jp,
-                                                    const JourneyPatternPoint& jpp)
-{
+        const JourneyPatternPoint& jpp)
```
Comment (pbougue) [inline source/routing/next_stop_time.cpp:57]: not a fan of that change (and following line)

File: source/kraken/tests/realtime_test.cpp [hunk 14]
```diff
@@ -738,8 +738,8 @@ BOOST_AUTO_TEST_CASE(train_delayed_day_after) {
     transit_realtime::TripUpdate trip_update = ntest::make_delay_message("vj:1",
             "20150928",
             {
-                    std::make_tuple("stop1", "20150929T0710"_pts, "20150929T0710"_pts),
```
Comment (antoine-de) [inline source/kraken/tests/realtime_test.cpp:741]: why did you need those changes ?

File: source/routing/next_stop_time.cpp (hunk 2 above)
Comment (xlqian) [inline source/routing/next_stop_time.cpp]: :+1:

File: source/routing/next_stop_time.cpp (hunk 3 above)
Comment (xlqian) [inline source/routing/next_stop_time.cpp]: ok

File: source/routing/next_stop_time.cpp (hunk 4 above)
Comment (xlqian) [inline source/routing/next_stop_time.cpp]: yep

File: source/routing/next_stop_time.cpp [hunk 15]
```diff
@@ -338,6 +341,126 @@ NextStopTime::tardiest_stop_time(const StopEvent stop_event,
     return first_discrete_st_pair;
 }
 
+
+template<typename F>
```
Comment (pbougue) [inline source/routing/next_stop_time.cpp:352]: a little comment on those 2 functions is welcome

File: source/routing/next_stop_time.cpp [hunk 16]
```diff
@@ -338,6 +341,126 @@ NextStopTime::tardiest_stop_time(const StopEvent stop_event,
...
+
+using DtSt = std::pair<DateTime, const type::StopTime*>;
+template<typename VJ_T>
+static void fill_cache(const DateTime from,
+        const DateTime to,
+        const type::RTLevel rt_level,
+        const type::AccessibiliteParams& accessibilite_params,
+        const JourneyPattern& jp,
+        const std::vector<const VJ_T*>& vjs,
+        IdxMap<JourneyPatternPoint, std::vector<DtSt>>& arrival,
```
Comment (pbougue) [inline source/routing/next_stop_time.cpp]: rename those params to know that it's "returned" : `arrival_cache` (or better if you find something else)

File: source/routing/next_stop_time.cpp [hunk 17]
```diff
@@ -338,6 +341,126 @@ NextStopTime::tardiest_stop_time(const StopEvent stop_event,
...
+}
+
+void CachedNextStopTime::load(const type::Data& data,
+        const DateTime from,
+        const DateTime to,
+        const type::RTLevel rt_level,
+        const type::AccessibiliteParams& accessibilite_params) {
+    const auto& jp_container = data.dataRaptor->jp_container;
+
+    departure.assign(jp_container.get_jpps_values());
```
Comment (pbougue) [inline source/routing/next_stop_time.cpp]: Those maps will be sparse... there might be a perf improvement in avoiding this, but I'm really not sure.
As this is critical, a bench on that may be a good idea. Any opinion @TeXitoi ?
We have to check what we do when no departure exists if we test.

File: source/routing/next_stop_time.cpp [hunk 18]
```diff
@@ -338,6 +341,126 @@ NextStopTime::tardiest_stop_time(const StopEvent stop_event,
...
+        boost::sort(jpp_dtst.second, compare);
+    }
+}
+
+std::pair<const type::StopTime*, DateTime>
+CachedNextStopTime::next_stop_time(const StopEvent stop_event,
+        const JppIdx jpp_idx,
+        const DateTime dt,
+        const bool clockwise) const {
+    const auto& v = stop_event == StopEvent::pick_up ? departure[jpp_idx] : arrival[jpp_idx];
```
Comment (pbougue) [inline source/routing/next_stop_time.cpp]: :sob:
`v` as name is short
add parenthesis?

File: source/routing/next_stop_time.cpp [hunk 19]
```diff
@@ -338,6 +341,126 @@ NextStopTime::tardiest_stop_time(const StopEvent stop_event,
...
+    const type::StopTime* null_st = nullptr;
+    decltype(v.begin()) search;
```
Comment (pbougue) [inline source/routing/next_stop_time.cpp:467]: a comment on that please?

File: source/routing/next_stop_time.cpp [hunk 20]
```diff
@@ -338,6 +341,126 @@ NextStopTime::tardiest_stop_time(const StopEvent stop_event,
...
+    if (clockwise) {
+        auto cmp = [](const DtSt& a, const DtSt& b) { return a.first < b.first; };
+        search = boost::lower_bound(v, std::make_pair(dt, null_st), cmp);
+    } else {
+        auto cmp = [](const DtSt& a, const DtSt& b) { return a.first < b.first; };
```
Comment (antoine-de) [inline source/routing/next_stop_time.cpp]: you can declare cmp above and use it no ?

File: source/routing/raptor.cpp [hunk 21]
```diff
@@ -41,6 +41,15 @@ namespace bt = boost::posix_time;
 
 namespace navitia { namespace routing {
 
+static DateTime limit_bound(const bool clockwise, const DateTime departure_datetime, const DateTime bound) {
+    return clockwise ?
+        std::min(departure_datetime + DateTimeUtils::SECONDS_PER_DAY, bound) :
+        std::max(departure_datetime > DateTimeUtils::SECONDS_PER_DAY ?
```
Comment (pbougue) [inline source/routing/raptor.cpp]: :sob: `max` + `?:`

File: source/routing/raptor.cpp [hunk 22]
```diff
@@ -172,21 +181,29 @@ void RAPTOR::init(const map_stop_point_duration& dep,
 void RAPTOR::first_raptor_loop(const map_stop_point_duration& dep,
                                const DateTime& departure_datetime,
                                const nt::RTLevel rt_level,
-                               const DateTime& bound,
+                               const DateTime& b,
```
Comment (pbougue) [inline source/routing/raptor.cpp:184]: another name for that... pliiiiize :)

File: source/routing/raptor.cpp [hunk 23]
```diff
@@ -472,16 +489,22 @@ RAPTOR::compute_nm_all(const std::vector<std::pair<type::EntryPoint, map_stop_po
                        const std::vector<std::pair<type::EntryPoint, map_stop_point_duration>>& arrivals,
                        const DateTime& departure_datetime,
                        const nt::RTLevel rt_level,
-                       const DateTime& bound,
+                       const DateTime& b,
```
Comment (pbougue) [inline source/routing/raptor.cpp]: same here :)

File: source/routing/raptor.cpp [hunk 24]
```diff
@@ -519,24 +542,31 @@ RAPTOR::compute_nm_all(const std::vector<std::pair<type::EntryPoint, map_stop_po
 void
 RAPTOR::isochrone(const map_stop_point_duration& departures,
                   const DateTime& departure_datetime,
-                  const DateTime& bound,
+                  const DateTime& b,
```
Comment (pbougue) [inline source/routing/raptor.cpp:488]: same here

File: source/routing/tests/raptor_test.cpp [hunk 25]
```diff
@@ -467,7 +469,7 @@ BOOST_AUTO_TEST_CASE(validity_pattern){
     BOOST_CHECK_EQUAL(res.items[0].arrival.time_of_day().total_seconds(), 9200);
 
     res1 = raptor.compute(d.stop_areas_map["stop1"], d.stop_areas_map["stop2"], 7800, 0, DateTimeUtils::set(0, 10000), type::RTLevel::Base, 2_min, true);
-    BOOST_REQUIRE_EQUAL(res1.size(), 1);
+    BOOST_REQUIRE_EQUAL(res1.size(), 1);//?
```
Comment (pbougue) [inline source/routing/tests/raptor_test.cpp]: remove that comment or add an explaining one?

File: source/routing/tests/raptor_test.cpp (hunk 10 above)
Comment (TeXitoi) [inline source/routing/tests/raptor_test.cpp:2048]: The test was buggy: we were getting the journey of the day before. Now that there is a 24h hard limit (else we have memory problems...), the test can't find the journey of the day before.

File: source/routing/next_stop_time.cpp (hunk 17 above)
Comment (TeXitoi) [inline source/routing/next_stop_time.cpp]: Those maps are IdxMap (that's why there is an empty vector on each entry), and have a O(1) access complexity. You can test with a flat_map or an unordered_map quite easily, but I'm septic on performance improvements.

Comment (pbougue) [inline source/routing/next_stop_time.cpp]: Ok, this does not prevent the PR to be merged, but a test with unordered_map might be nice if we have time for it.

Comment (xlqian) [inline source/routing/next_stop_time.cpp]: to @all, I just finished my ticket, I can do a bench on that (as I already have all settings for that.)

**Comment by xlqian** [comment]
Result of bench:
* 100 requests, every request is repeated for 10 times
* Benched on BackupStif, fr-pdl, SNCF
* Benched on `dev` (right before this PR, contributor is only activated on VJ), `Cache`, `LRU`(to do)

**Backup STIF** :

| (nb 2nd pass) 	| 0      	| 10      	| 100     	| 1000    	| 10000   	|
|---------------	|--------	|---------	|---------	|---------	|---------	|
| dev           	| 49.81s 	| 96.56s  	| 290.97s 	| 347.7s  	| 366.08s 	|
| Cache         	| 83.3s  	| 100.36s 	| 183.79s 	| 204.49s 	| 223.35s 	|
| LRU           	|        	|         	|         	|         	|         	|


**fr-pdl** 

| (nb 2nd pass) 	| 0         	| 10     	| 100    	| 1000 	| 10000 	|
|---------------	|-----------	|--------	|--------	|------	|-------	|
| dev           	| 78.878.8s 	| 128s   	| 251    	| 280s 	| 281s  	|
| Cache         	| 120s      	| 126.13 	| 193.09 	| 203s 	| 204s  	|
| LRU           	|           	|        	|        	|      	|       	|

**SNCF**

| (nb 2nd pass) 	| 0      	| 10     	| 100    	| 1000   	| 10000  	|
|---------------	|--------	|--------	|--------	|--------	|--------	|
| dev           	| 11.62s 	| 12.37s 	| 13.17s 	| 12.36s 	| 12.38s 	|
| Cache         	| 15.7s  	| 16.01  	| 16.17s 	| 16.73s 	| 16.69s 	|
| LRU           	|        	|        	|        	|        	|        	|

**Comment by xlqian** [comment]
Result of bench:
* 100 requests, every request is repeated for 10 times
* Benched on BackupStif, fr-pdl, SNCF
* Benched on `dev` (right before this PR, contributor is only activated on VJ), `Cache`, `LRU`(to do)

**Backup STIF** :

| (nb 2nd pass) 	| 0      	| 10      	| 100     	| 1000    	| 10000   	|
|---------------	|--------	|---------	|---------	|---------	|---------	|
| dev           	| 49.81s 	| 96.56s  	| 290.97s 	| 347.7s  	| 366.08s 	|
| Cache         	| 83.3s  	| 100.36s 	| 183.79s 	| 204.49s 	| 223.35s 	|
| LRU           	|        	|         	|         	|         	|         	|


**fr-pdl** 

| (nb 2nd pass) 	| 0         	| 10     	| 100    	| 1000 	| 10000 	|
|---------------	|-----------	|--------	|--------	|------	|-------	|
| dev           	| 78.878.8s 	| 128s   	| 251    	| 280s 	| 281s  	|
| Cache         	| 120s      	| 126.13 	| 193.09 	| 203s 	| 204s  	|
| LRU           	|           	|        	|        	|      	|       	|

**SNCF**

| (nb 2nd pass) 	| 0      	| 10     	| 100    	| 1000   	| 10000  	|
|---------------	|--------	|--------	|--------	|--------	|--------	|
| dev           	| 11.62s 	| 12.37s 	| 13.17s 	| 12.36s 	| 12.38s 	|
| Cache         	| 15.7s  	| 16.01  	| 16.17s 	| 16.73s 	| 16.69s 	|
| LRU           	|        	|        	|        	|        	|        	|

File: source/routing/next_stop_time.cpp (hunk 5 above)
Comment (xlqian) [inline source/routing/next_stop_time.cpp]: done

File: source/routing/next_stop_time.cpp (hunk 6 above)
Comment (xlqian) [inline source/routing/next_stop_time.cpp]: done

File: source/routing/next_stop_time.cpp (hunk 7 above)
Comment (xlqian) [inline source/routing/next_stop_time.cpp:368]: Ok, I benched with fr-pdl

File: source/routing/next_stop_time.h (hunk 8 above)
Comment (xlqian) [inline source/routing/next_stop_time.h]: done

File: source/routing/benchmark_full.cpp (hunk 12 above)
Comment (xlqian) [inline source/routing/benchmark_full.cpp]: done

File: source/routing/next_stop_time.cpp (hunk 13 above)
Comment (xlqian) [inline source/routing/next_stop_time.cpp:57]: sorry, IDE indented for me

File: source/routing/next_stop_time.cpp (hunk 15 above)
Comment (xlqian) [inline source/routing/next_stop_time.cpp:352]: yep

File: source/routing/next_stop_time.cpp (hunk 17 above)
Comment (xlqian) [inline source/routing/next_stop_time.cpp]: ok, maybe next time :)

File: source/routing/next_stop_time.cpp (hunk 20 above)
Comment (xlqian) [inline source/routing/next_stop_time.cpp]: done

File: source/routing/tests/raptor_test.cpp (hunk 25 above)
Comment (xlqian) [inline source/routing/tests/raptor_test.cpp]: yep

File: source/kraken/tests/realtime_test.cpp (hunk 14 above)
Comment (xlqian) [inline source/kraken/tests/realtime_test.cpp:741]: same explanation as above :)

**Comment by pbougue** [comment]
PR done on your branch for LRU: It's possible to bench (just missing cache size param?)

**Comment by pbougue** [comment]
btw, any return on those hard request for BackupStif?

File: source/kraken/configuration.cpp [hunk 26]
```diff
@@ -174,4 +179,15 @@ bool Configuration::display_contributors() const{
     }
     return vm["GENERAL.display_contributors"].as<bool>();
 }
+
+size_t Configuration::raptor_cache_size() const{
+    if (! vm.count("GENERAL.raptor_cache_size")) {
```
Comment (TeXitoi) [inline source/kraken/configuration.cpp:184]: There is already a default value that should handle that.

File: source/kraken/configuration.cpp [hunk 27]
```diff
@@ -118,8 +119,12 @@ boost::optional<std::string> Configuration::chaos_database() const{
...
     return result;
 }
-int Configuration::nb_thread() const{
-    return this->vm["GENERAL.nb_threads"].as<int>();
+int Configuration::nb_threads() const{
+    int nb_threads = vm["GENERAL.nb_threads"].as<int>();
+    if (nb_threads < 0) {
+        throw std::invalid_argument("nb_threads cannot be negative");
+    }
+    return size_t(nb_threads);
```
Comment (TeXitoi) [inline source/kraken/configuration.cpp:127]: Why?

File: source/routing/next_stop_time.cpp [hunk 28]
```diff
@@ -1,38 +1,41 @@
...
 IRC #navitia on freenode
 https://groups.google.com/d/forum/navitia
 www.navitia.io
-*/
+ */
 
 #include "next_stop_time.h"
 
 #include "dataraptor.h"
+#include "get_stop_times.h"
```
Comment (TeXitoi) [inline source/routing/next_stop_time.cpp]: I think this include is no more needed.

File: source/routing/raptor.cpp (hunk 22 above)
Comment (TeXitoi) [inline source/routing/raptor.cpp:184]: There is an explicit name 5 lines after.

File: source/routing/raptor.cpp (hunk 23 above)
Comment (TeXitoi) [inline source/routing/raptor.cpp]: same here ;-)

File: source/routing/raptor.cpp [hunk 29]
```diff
@@ -472,16 +487,20 @@ RAPTOR::compute_nm_all(const std::vector<std::pair<type::EntryPoint, map_stop_po
...
                        const type::AccessibiliteParams& accessibilite_params,
                        const std::vector<std::string>& forbidden_uri,
                        bool clockwise) {
     std::vector<std::pair<type::EntryPoint, std::vector<Path>>> result;
-    set_valid_jp_and_jpp(DateTimeUtils::date(departure_datetime),
-                         accessibilite_params,
-                         forbidden_uri,
-                         rt_level);
+    const DateTime bound = limit_bound(clockwise, departure_datetime, b);
+    const auto valid_jpps = set_valid_jp_and_jpp(DateTimeUtils::date(departure_datetime),
```
Comment (TeXitoi) [inline source/routing/raptor.cpp]: valid_jpps doesn't seems used.

File: source/routing/raptor.cpp [hunk 30]
```diff
@@ -519,24 +538,29 @@ RAPTOR::compute_nm_all(const std::vector<std::pair<type::EntryPoint, map_stop_po
...
                   const type::AccessibiliteParams& accessibilite_params,
                   const std::vector<std::string>& forbidden,
                   bool clockwise,
                   const nt::RTLevel rt_level) {
-    set_valid_jp_and_jpp(DateTimeUtils::date(departure_datetime),
-                         accessibilite_params,
-                         forbidden,
-                         rt_level);
+    const DateTime bound = limit_bound(clockwise, departure_datetime, b);
+    const auto valid_jpps = set_valid_jp_and_jpp(DateTimeUtils::date(departure_datetime),
```
Comment (TeXitoi) [inline source/routing/raptor.cpp]: valid_jpps doesn't seems used.

File: source/routing/raptor.cpp [hunk 31]
```diff
@@ -519,24 +538,29 @@ RAPTOR::compute_nm_all(const std::vector<std::pair<type::EntryPoint, map_stop_po
...
     init(departures, departure_datetime, clockwise, accessibilite_params.properties);
 
-    boucleRAPTOR(accessibilite_params, clockwise, rt_level, max_transfers);
+    boucleRAPTOR(clockwise, rt_level, max_transfers);
 }
 
-
-void RAPTOR::set_valid_jp_and_jpp(
+// Returns valid_jpps
+boost::dynamic_bitset<> RAPTOR::set_valid_jp_and_jpp(
```
Comment (TeXitoi) [inline source/routing/raptor.cpp]: now useless?

File: source/routing/tests/routing_api_test.cpp (hunk 11 above)
Comment (TeXitoi) [inline source/routing/tests/routing_api_test.cpp:1296]: Because of the now hard limit of 24h

**Comment by TeXitoi** [comment]
:+1: great job guys!

File: source/kraken/configuration.cpp (hunk 26 above)
Comment (antoine-de) [inline source/kraken/configuration.cpp:184]: no it's mandatory to check

File: source/kraken/configuration.cpp (hunk 27 above)
Comment (antoine-de) [inline source/kraken/configuration.cpp:127]: I do agree, we should return a size_t

File: source/kraken/configuration.cpp (hunk 26 above)
Comment (pbougue) [inline source/kraken/configuration.cpp:184]: If you don't read a file you don't have value at all

File: source/kraken/configuration.cpp (hunk 27 above)
Comment (pbougue) [inline source/kraken/configuration.cpp:127]: sorry for that, we have to remove cast to be consistent (how we use nb_threads() now is safe)
and as it was not the point of the PR I didn't want to refactor the int on all use

File: source/routing/raptor.cpp (hunk 21 above)
Comment (xlqian) [inline source/routing/raptor.cpp]: done

File: source/routing/next_stop_time.cpp (hunk 28 above)
Comment (xlqian) [inline source/routing/next_stop_time.cpp]: ok

File: source/routing/raptor.cpp (hunk 30 above)
Comment (xlqian) [inline source/routing/raptor.cpp]: ok

**Comment by xlqian** [comment]
Jenkins, test this please!
